### PR TITLE
[features] bring your own vectors

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
we allow the customer to bring their own vectors in the search

* **What is the current behavior?** (You can also link to an open issue here)
no custom vectors
* **What is the new behavior (if this is a feature change)?**
they can bring own vectors

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
running

* **Related Python client changes** (link commit/PR here)
will do

* **Related documentation changes** (link commit/PR here)
will do

* **Other information**:
the [design doc](https://docs.google.com/document/d/1gUB69UNj7mq-YW4sbi7fqwwiHRyOHkLQ-3dTzP2iJ4c/edit)

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/tensor_search/validation.py [hunk 1]
```diff
@@ -52,6 +53,9 @@ def validate_query(q: Union[dict, str], search_method: Union[str, SearchMethod])
     return q
 
 
+def validate_context(context:dict=None):
+    pass
```
Comment (pandu-k) [inline src/marqo/tensor_search/validation.py]: What is this?

Comment (wanliAlex) [reply src/marqo/tensor_search/validation.py]: forget to delete! revised

File: src/marqo/tensor_search/tensor_search.py [hunk 2]
```diff
@@ -1278,13 +1282,19 @@ def _vector_text_search(
...
+        validation.validate_context_object(context_object=context)
+        custom_tensors = context.get("tensor", None)
     start_preprocess_time = timer()
     try:
         index_info = get_index_info(config=config, index_name=index_name)
     except KeyError as e:
         raise errors.IndexNotFoundError(message="Tried to search a non-existent index: {}".format(index_name))
     selected_device = config.indexing_device if device is None else device
 
+    print(index_info.index_settings["index_defaults"]["model"])
```
Comment (Jeadie) [inline src/marqo/tensor_search/tensor_search.py]: Do we want this in long-term?

Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py]: sorry. forget to delete. solved

**Review by wanliAlex — COMMENTED** [review]
updated

**Review by Jeadie — APPROVED** [review]
Good work. I think this implementation brings up two good questions we should consider in open-source team. 
1. Do we want to let basic queries (e.g. `q="standard string"`) have an associated context vector?
2. We can reduce search complexity by converting `q="standard string"` into `q={"standard string": 1.0}` at the very start.

**Review by Jeadie — APPROVED** [review]

File: tests/tensor_search/test_custom_vectors_search.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                "_id": "1"
+            }], auto_refresh=True)
+
+    def tearDown(self) -> None:
+        try:
+            tensor_search.delete_index(config=self.config, index_name=self.index_name_1)
+        except:
+            pass
+
+    def test_search(self):
```
Comment (Jeadie) [inline tests/tensor_search/test_custom_vectors_search.py:47]: We aren't really testing anything here?

Comment (wanliAlex) [reply tests/tensor_search/test_custom_vectors_search.py:47]: just to test the feature.

**Comment by wanliAlex** [comment]
For manual test,
```python        
query = {
            "A rider is riding a horse jumping over the barrier": 0,
        }

res_2 = tensor_search.search(config=self.config, index_name=self.index_name_1, text=query, context=
{"tensor": [{"vector": [3, ] * 512, "weight": 1}, {"vector": [2, ] * 512, "weight": 0}], })
```

this search generates the search vector (before normalization):
`[1. , 1., ... , 1.]` which is expected.

**Pull request merged**